Add SVSM e-vTPM support to go-tpm-tools

jxu023 opened the pull request:
SVSM e-vTPM currently can only attest to an EK, so we need EK-based key attestation for the TPM.

This PR implements https://trustedcomputinggroup.org/wp-content/uploads/EK-Based-Key-Attestation-with-TPM-Firmware-Version-V1-RC1_9July2025.pdf#page=8 and is based on the reference code https://github.com/TrustedComputingGroup/tpm-fw-attestation-reference-code

chrisfenner commented on client/import_certify.go:
```diff
@@ -0,0 +1,140 @@
+
+// SolveChallengeImportCertify solves a challenge from the TPM EK.
+func SolveChallengeImportCertify(tpm transport.TPM, req *tpb.ImportBlob, keyAlgo legacytpm2.Algorithm) (*rpb.CertifiedBlob, error) {
+	// SVSM currently only supports attesting an RSA EK.
```
I would recommend supporting using the ECC EK (e.g., P256) sooner than later. It's much faster (esp. key generation) and a little bit more secure (trade-off is win/win).

jxu023 replied:
Created a bug to do this in backlog

chrisfenner commented on server/import_certify.go:
```diff
@@ -0,0 +1,188 @@
+		return fmt.Errorf("%w: unexpected prefix %0x", errInvalidAttestation, attest.Magic)
+	}
+
+	// TODO: check qualified signer?
```
I don't think we need to check the qualified signer, since the signer must be the same HMAC key we put into the challenge blob in order for the HMAC verify to pass

chrisfenner reviewed (approved):
No serious complaints from my end, although I focused my review on the crypto & attestation flow rather than the proto design.

jxu023 commented on client/import_certify.go:
```diff
@@ -0,0 +1,140 @@
+import (
+	"fmt"
+
+	legacytpm2 "github.com/google/go-tpm/legacy/tpm2"
```
Updated to use tpm2.TPMAlgID around these files, cmd/register.go now converts the uint16 to TPMAlgID (since we're using the existing keyAlgo flag shared with other commands).

jxu023 commented on server/svsm.go:
```diff
@@ -0,0 +1,43 @@
+	rpb "github.com/google/go-tpm-tools/proto/register_credential"
+)
+
+// VerifySvsmAttestation checks the SNP attestation report to ensure it was requested by
```
Done

alexmwu commented on cmd/attest.go:
```diff
@@ -51,6 +54,7 @@ key or AK for a custom attestation key. By default it uses AK.
 by default rsa is used.
 --tee-nonce attaches a 64 bytes extra data to the attestation report of TDX and SEV-SNP 
 hardware and guarantees a fresh quote.
+--svsm will specify svsm as the service provider (Linux only) for the attestation report.
```
--svsm will specify SVSM as the...

alexmwu commented on client/svsm.go:
```diff
@@ -0,0 +1,93 @@
+func MakeSVSMAttestation(tpm transport.TPM, req *tpb.ImportBlob, keyAlgo tpm2.TPMAlgID, teeNonce [sabi.ReportDataSize]byte) (*rpb.SevSnpSvsmAttestation, error) {
+	certified, err := SolveChallengeImportCertify(tpm, req, keyAlgo)
+	if err != nil {
+		return nil, fmt.Errorf("SolveChallengeImportCertify() = %w", err)
```
here and elsewhere, please match the error string style. for example: "failed to create import certify solution".

alexmwu commented on cmd/pubkey.go:
```diff
@@ -12,8 +12,15 @@ import (
+var keyFormat string
+
+func addKeyFormatFlag(cmd *cobra.Command) {
+       cmd.PersistentFlags().StringVar(&keyFormat, "key-format", "pem", "type of format for the outputted key, defaults to pem, but can also specify tpm2-wire-format")
```
Be specific about the format (TPMT_PUBLIC/tpmt-public)

alexmwu commented on cmd/register.go:
```diff
@@ -0,0 +1,147 @@
+	"github.com/google/go-tpm-tools/server"
+)
+
+// this file implements TPM registration which for now is only used for SVSM e-vTPMs.
```
Capitalize comments and end with periods

alexmwu commented on server/svsm.go:
```diff
@@ -0,0 +1,43 @@
+// SVSM from VMPL0 and that report data contains a hash of the tee nonce and vTPM service manifest.
+// other SNP attestation report values (including the launch measurement)
+// should be checked via gce tcb verifier's sev validate command
+func VerifySVSMAttestation(teeNonce [sevabi.ReportDataSize]byte, attestation *rpb.SevSnpSvsmAttestation, secret []byte) error {
```
VerifySVSMAttestation(attestation *rpb.SevSnpSvsmAttestation, opts VerifySVSMOpts)

alexmwu commented on server/import_certify.go:
```diff
@@ -0,0 +1,189 @@
+}
+
+// VerifyChallenge verifies the results of the TPM EK registration flow.
+func VerifyChallenge(req *rpb.CertifiedBlob, secret []byte) error {
```
Let's avoid exporting unless we need to. I'd prefer to export when we need it rather than intro breaking changes or unexport.

jxu023 replied:
unexported

alexmwu commented on cmd/attest.go:
```diff
@@ -121,16 +121,21 @@ hardware and guarantees a fresh quote.
-			}
-		} else {
-			out = []byte(marshalOptions.Format(attestation))
+		if svsm != "" {
```
To be clear, we require the person to run the go-sev-guest CLI first and attach it to this proto via the --svsm flag? Wouldn't it be more usable to just create an appropriate SNP attestation in line here?

jxu023 replied:
created svsm subcommand to attach the SVSM attestation info

alexmwu reviewed (commented):
Can we add tests for these new additions? I remember go-sev-guest supports a mock https://github.com/google/go-sev-guest/blob/main/testing/test_cases.go#L304 but I'm not too familiar with it tbh

I'll need to do another round on the library crypto; thankfully Chris has already taken a look.

jxu023 commented on client/svsm.go:
```diff
@@ -0,0 +1,93 @@
+)
+
+// MakeSVSMAttestation creates a SevSnpSvsmAttestation containing all the information needed to verify the SVSM e-vTPM.
+func MakeSVSMAttestation(tpm transport.TPM, req *tpb.ImportBlob, keyAlgo tpm2.TPMAlgID, teeNonce [sabi.ReportDataSize]byte) (*rpb.SevSnpSvsmAttestation, error) {
```
Done.

jxu023 commented on proto/register_credential.proto:
```diff
@@ -0,0 +1,27 @@
+
+// Contains proof that the ak_pub is co-resident with an EK
+message CertifiedBlob {
+  // TPM2BPublic not including length field
```
Done

jxu023 commented on proto/register_credential.proto:
```diff
@@ -0,0 +1,27 @@
+  sevsnp.Attestation sev_snp_attestation = 1;
+  CertifiedBlob certified_blob = 2;
+  // vTPM service manifest retrieved from configfs-tsm via manifest blob
+  // its format is defined by the SVSM spec
```
Done

jxu023 commented on client/import_certify.go:
```diff
@@ -0,0 +1,139 @@
+	}.Execute(tpm)
+
+	// Certify a newly created AK
+	certified, err := tpm2.Certify{
```
per the EK based remote attestation w/ firmware spec, we can use a nonce with CertifyCreation(); I'll wait on chris to respond on the necessity. 

I suppose adding a nonce means we should also called the TPM clear() operation? Otherwise re-trying this section of code on the same TPM will fail.

jxu023 commented on client/svsm.go:
```diff
@@ -0,0 +1,93 @@
+)
+
+// MakeSVSMAttestation creates a SevSnpSvsmAttestation containing all the information needed to verify the SVSM e-vTPM.
+func MakeSVSMAttestation(tpm transport.TPM, req *tpb.ImportBlob, keyAlgo tpm2.TPMAlgID, teeNonce [sabi.ReportDataSize]byte) (*rpb.SevSnpSvsmAttestation, error) {
```
Done.

jxu023 commented on server/import_certify.go:
```diff
@@ -0,0 +1,189 @@
+)
+
+// CreateChallenge generates a new HMAC key and wraps it to the given EK.
+func CreateChallenge(tPublic *tpm2.TPMTPublic) (*tpb.ImportBlob, []byte, error) {
```
Done.

jxu023 commented:
> Can we add tests for these new additions? I remember go-sev-guest supports a mock https://github.com/google/go-sev-guest/blob/main/testing/test_cases.go#L304 but I'm not too familiar with it tbh
> 
> I'll need to do another round on the library crypto; thankfully Chris has already taken a look.

I've uploaded another revision, still working on unit tests. I'll be updating the SVSMOpts to include configfsi and LeveledQuoteProvider interfaces so that I can inject a test double there

jxu023 commented:
Updated with unit tests

alexmwu commented on server/import_certify.go:
```diff
@@ -0,0 +1,188 @@
+	"github.com/google/go-tpm/tpm2"
+)
+
+// this file aims to implement the verifier side of https://trustedcomputinggroup.org/wp-content/uploads/EK-Based-Key-Attestation-with-TPM-Firmware-Version-V1-RC1_9July2025.pdf#page=8
```
Capitalize comments.

alexmwu commented on server/import_certify.go:
```diff
@@ -0,0 +1,188 @@
+	}, hmacKey, nil
+}
+
+func VerifyCertifiedAKBlob(req *rpb.CertifiedBlob, secret []byte) error {
```
This should probably be unexported if we're not exporting the client func

jxu023 replied:
exported since now gotpm register cmd uses this instead of the SVSM wrapper

alexmwu commented on server/import_certify.go:
```diff
@@ -0,0 +1,188 @@
+// for reference: https://github.com/TrustedComputingGroup/tpm-fw-attestation-reference-code
+
+var (
+	errCertifiedWrongName = errors.New("incorrect name")
```
nit: Name or TPM name

alexmwu commented on server/import_certify.go:
```diff
@@ -0,0 +1,188 @@
+	h.Write(obfuscate)
+	h.Write(hmacKey[:])
+
+	pub := &tpm2.TPMTPublic{
```
This is heavily duplicated with https://github.com/google/go-tpm-tools/blob/main/server/key_conversion.go#L52-L64. Can we extend that to pass in the correct Object attributes (sealedDataObjects leave it unset; imported HMAC keys should set)

alexmwu commented on server/import_certify.go:
```diff
@@ -0,0 +1,188 @@
+		}),
+	}
+
+	priv := &tpm2.TPMTSensitive{
```
duplicated with https://github.com/google/go-tpm-tools/blob/main/server/key_conversion.go#L66-L77

alexmwu commented on server/import_certify.go:
```diff
@@ -0,0 +1,188 @@
+}
+
+// verifyCertify checks the certifyInfo against the given name.
+func verifyCertify(name *tpm2.TPM2BName, certifyInfo *tpm2.TPMSCertifyInfo) error {
```
verifyCertifyInfo

rossjudson commented on client/svsm.go:
```diff
@@ -0,0 +1,109 @@
+	}
+	reportWithCerts, err := sabi.ReportCertsToProto(rawWithCerts)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse SEV-SNP report with certs: %v", err)
```
sabi.ReportsCertsToProto uses the %v err formatter, dropping context. Elsewhere, sabi.ReportToProto uses %v -- not dropping context.

Is this inconsistency intentional?

jxu023 replied:
TIL about %w vs %v. I've replaced usages with %w in new code, except in unit tests where t.Fatalf() doesn't support %w.

alexmwu commented on server/svsm.go:
```diff
@@ -0,0 +1,51 @@
+		return fmt.Errorf("challenge verification failed: %w", err)
+	}
+
+	h := sha512.New()
```
It would be useful to link the spec or the code for this protocol.

jxu023 replied:
added spec and code links

alexmwu commented on server/svsm.go:
```diff
@@ -0,0 +1,51 @@
+		return fmt.Errorf("report data does not match expected value")
+	}
+	if report.Vmpl != 0 {
+		return fmt.Errorf("attestation report was not requested from VMPL0, SVSM should be in VMPL0")
```
nit: this can be a sentinel error.

jkl73 commented on client/import_certify.go:
```diff
@@ -0,0 +1,141 @@
+	if err != nil {
+		return nil, fmt.Errorf("failed to create RSA EK: %w", err)
+	}
+	defer tpm2.FlushContext{
```
looks like this is already flushed below line 95

jxu023 replied:
this was added to ensure flushing from the err conditions despite the duplicate flush.

added explicit flush from the err cond early returns instead.

jkl73 commented on client/import_certify.go:
```diff
@@ -0,0 +1,141 @@
+		ParentHandle: tpm2.AuthHandle{
+			Handle: ek.ObjectHandle,
+			Name:   ek.Name,
+			Auth:   tpm2.Policy(tpm2.TPMAlgSHA256, 16, ekPolicy),
```
I thought the auth size is suppose to be 32 instead of 16? (I might be wrong)

jxu023 replied:
I tried with 32 and 16, and they both seem to work: from checking the TPM library ref: https://trustedcomputinggroup.org/wp-content/uploads/Trusted-Platform-Module-2.0-Library-Part-1-Version-184_pub.pdf

It notes in _17.6.3.2 Session Nonce Size_
> The minimum size for nonceCaller in TPM2_StartAuthSession() is 16 octets
> ... The maximum size that may be requested for nonceTPM is the size of the digest produced by the authorization session hash
> For SHA-1, the maximum size for nonceTPM is 20 octets and for SHA256 it is 32 octets.

Octet means byte apparently.

i think both work. I've updated the value to 32 cuz why not, more secure probably.

jkl73 commented on cmd/register.go:
```diff
@@ -0,0 +1,158 @@
+		challenge := &tpb.ImportBlob{}
+		err := readProtoFromPath(input, challenge)
+		if err != nil {
+			return fmt.Errorf("could not read challnge: %s", err)
```
typo: challenge

jkl73 reviewed (approved):
lgtm two overall thoughts: 
* The cmd looks a bit complicated, we might think how to streamline it.
* svsm specific code in client/server might be more suitable to put into go-sev-guest repo

jxu023 commented on client/import_certify.go:
```diff
@@ -0,0 +1,141 @@
+func ekPub(tpm transport.TPM) ([]byte, error) {
+	cp, err := tpm2.CreatePrimary{
+		PrimaryHandle: tpm2.TPMRHEndorsement,
+		InPublic:      tpm2.New2B(tpm2.RSAEKTemplate),
```
I used alex's suggestion and created an ekResponse() function wrapper.

jxu023 commented on client/import_certify.go:
```diff
@@ -0,0 +1,141 @@
+	}
+	if keyAlgo == tpm2.TPMAlgRSA {
+		public, err = AKTemplateRSA().Encode()
+	}
```
modified to switch and added error on unsupported keyAlgo

jxu023 commented on cmd/register.go:
```diff
@@ -0,0 +1,158 @@
+
+func init() {
+	RootCmd.AddCommand(registerCmd)
+	challengeCommands := []*cobra.Command{
```
Moved svsm logic to svsm subcommand under attest and verify, register.go is now SVSM-free (except for one comment noting that we're only creating an EK RSA because that's all that's currently supported on SVSM).

jxu023 commented on server/import_certify.go:
```diff
@@ -0,0 +1,188 @@
+}
+
+// verifyAttest checks that the attestation structure has valid data
+func verifyAttest(attest *tpm2.TPMSAttest) error {
```
updated name

jxu023 commented on server/svsm.go:
```diff
@@ -0,0 +1,51 @@
+
+type VerifySVSMOpts struct {
+	TEENonce    []byte
+	Attestation *rpb.SevSnpSvsmAttestation
```
moved to top-level param

jxu023 commented on server/svsm.go:
```diff
@@ -0,0 +1,51 @@
+	Attestation *rpb.SevSnpSvsmAttestation
+	Secret      []byte
+	VerifyOpts  *verify.Options
+}
```
passed in validate options, and also attached endorsement into the SVSM message. Also added verification on endorsement by adding gcetcbendorsement as a dependency.

Now we no longer need to make separate CLI call to gce-tcb-verifier/gce-tcb-endorsement!

jxu023 commented:
> The cmd looks a bit complicated, we might think how to streamline it.

I've removed SVSM logic from tpm registration, so hopefully it's a bit simpler now.

> svsm specific code in client/server might be more suitable to put into go-sev-guest repo

I've moved svsm specific code to the cmd package instead of client/server. This way we can do a drop-in replacement when i move some of that to either go-sev-guest or tcb-verifier

I've uploaded a new revision which includes these changes:
* added go-sev-guest v0.14.0 to deps so that i can abandon https://github.com/google/go-tpm-tools/pull/608 if we merge this
*  added 2 second sleep prior to SVSM attestation report request (since Coconut SVSM will crash on hitting the guest request ratelimit)
* made svsm subcommands for gotpm attest/verify
* added gce-tcb-endorsement attestation/verification into svsm subcommands

@alexmwu @jkl73 lemme know how the changes look

Notably we no longer need to separately call gce-tcb-verifier CLI
full example usage:
```
TEE_NONCE="123456789012345678901234567890123456789012345678901234567890123456789012345
"
ALGO="ecc"

"${gotpm}" pubkey endorsement --key-format=tpmt-public \
  --algo=rsa --output=ek.bin
"${gotpm}" register create-challenge --input=ek.bin \
  --output=challenge.bin --secret-output=secret.bin
"${gotpm}" register solve-challenge --input=challenge.bin --algo=${ALGO} \
  --output=certified_blob.bin
"${gotpm}" register verify-challenge --input=certified_blob.bin \
  --secret-input=secret.bin

"${gotpm}" attest svsm --nonce=1234 --output=svsm_attestation.bin \
  --algo=${ALGO} --key=AK --tee-nonce="${TEE_NONCE}" \
  --tee-technology=sev-snp
"${gotpm}" verify debug svsm --nonce=1234 --input=svsm_attestation.bin \
  --tee-technology=sev-snp --certified-ak-blob=certified_blob.bin \
  --ek-pub=ek.bin --tee-nonce="${TEE_NONCE}"
```

mayafleischer commented:
/gcbrun

mayafleischer commented:
/gcbrun

alexmwu commented on client/import_certify_test.go:
```diff
@@ -0,0 +1,257 @@
+		{"RSA", tpm2.TPMAlgRSA},
+		{"ECC", tpm2.TPMAlgECC},
+	}
+	for _, tc := range testcases {
```
nit: could the CreateCertifiedAKBlob be a receiver method on the ek Key. You can then call ek.PublicArea() below and ek.CreateCertifiedAKBlob(tpm, challenge, tc.keyAlgo)

alexmwu commented on server/import_certify.go:
```diff
@@ -0,0 +1,188 @@
+// for reference: https://github.com/TrustedComputingGroup/tpm-fw-attestation-reference-code
+
+var (
+	errCertifiedWrongName = errors.New("incorrect name")
```
"incorrect TPM name"
